File: initial_exploration/sample_repo/csharp/sample.cs
/**
 * Sample C# file for tree-sitter exploration.
 * Includes modern C# constructs.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Sample
{
    // Record type (C# 9+)
    public record UserRecord(string Id, string Name, string? Email = null);

    // Enum
    public enum Status
    {
        Pending,
        Active,
        Inactive
    }

    // Interface with default implementation
    public interface IRepository<T> where T : class
    {
        Task<T?> FindAsync(string id);
        Task SaveAsync(T item);
        Task<bool> DeleteAsync(string id);

        // Default implementation
        async Task<IEnumerable<T>> FindAllAsync(IEnumerable<string> ids)
        {
            var results = new List<T>();
            foreach (var id in ids)
            {
                var item = await FindAsync(id);
                if (item != null) results.Add(item);
            }
            return results;
        }
    }

    // Class with various features
    public class User : IEquatable<User>
    {
        // Auto-properties
        public string Id { get; init; }
        public string Name { get; set; }
        public string? Email { get; set; }
        public Status Status { get; private set; } = Status.Pending;

        // Expression-bodied property
        public bool IsActive => Status == Status.Active;

        // Constructor
        public User(string id, string name)
        {
            Id = id ?? throw new ArgumentNullException(nameof(id));
            Name = name ?? throw new ArgumentNullException(nameof(name));
        }

        // Method with expression body
        public void Activate() => Status = Status.Active;

        // Async method
        public async Task<string> FetchDataAsync(string url)
        {
            await Task.Delay(100);
            return $"Data from {url}";
        }

        // Pattern matching
        public string GetStatusMessage() => Status switch
        {
            Status.Pending => "Waiting for activation",
            Status.Active => "User is active",
            Status.Inactive => "User has been deactivated",
            _ => throw new InvalidOperationException()
        };

        // IEquatable implementation
        public bool Equals(User? other)
        {
            if (other is null) return false;
            return Id == other.Id;
        }

        public override bool Equals(object? obj) => Equals(obj as User);
        public override int GetHashCode() => Id.GetHashCode();

        // Deconstruction
        public void Deconstruct(out string id, out string name)
        {
            id = Id;
            name = Name;
        }
    }

    // Generic class with constraints
    public class InMemoryRepository<T> : IRepository<T> where T : class
    {
        private readonly Dictionary<string, T> _items = new();
        private readonly Func<T, string> _keySelector;

        public InMemoryRepository(Func<T, string> keySelector)
        {
            _keySelector = keySelector;
        }

        public Task<T?> FindAsync(string id)
        {
            _items.TryGetValue(id, out var item);
            return Task.FromResult(item);
        }

        public Task SaveAsync(T item)
        {
            var key = _keySelector(item);
            _items[key] = item;
            return Task.CompletedTask;
        }

        public Task<bool> DeleteAsync(string id)
        {
            return Task.FromResult(_items.Remove(id));
        }
    }

    // Abstract class
    public abstract class BaseService<T> where T : class
    {
        protected readonly IRepository<T> Repository;

        protected BaseService(IRepository<T> repository)
        {
            Repository = repository;
        }

        public abstract Task<bool> ValidateAsync(T item);

        public virtual async Task<T?> GetByIdAsync(string id)
        {
            return await Repository.FindAsync(id);
        }
    }

    // Inheritance
    public class UserService : BaseService<User>
    {
        public UserService(IRepository<User> repository) : base(repository) { }

        public override Task<bool> ValidateAsync(User item)
        {
            return Task.FromResult(!string.IsNullOrEmpty(item.Name));
        }

        // LINQ example
        public async Task<IEnumerable<User>> GetActiveUsersAsync(IEnumerable<string> ids)
        {
            var users = await Repository.FindAllAsync(ids);
            return users.Where(u => u.IsActive)
                       .OrderBy(u => u.Name)
                       .ToList();
        }
    }

    // Extension methods
    public static class UserExtensions
    {
        public static string ToDisplayString(this User user)
        {
            return $"{user.Name} ({user.Id})";
        }

        public static async Task<User> WithEmailAsync(this User user, string email)
        {
            await Task.Delay(10); // Simulate async operation
            user.Email = email;
            return user;
        }
    }

    // Attribute
    [AttributeUsage(AttributeTargets.Method)]
    public class LogAttribute : Attribute
    {
        public string Message { get; }
        public LogAttribute(string message) => Message = message;
    }

    // Static class
    public static class Validators
    {
        public static bool IsValidEmail(string? email)
            => !string.IsNullOrEmpty(email) && email.Contains('@');

        public static bool IsNotEmpty(string? value)
            => !string.IsNullOrWhiteSpace(value);
    }

    // Partial class
    public partial class PartialUser
    {
        public string Id { get; set; } = string.Empty;
    }

    public partial class PartialUser
    {
        public string Name { get; set; } = string.Empty;
    }

    // Main program
    public class Program
    {
        public static async Task Main(string[] args)
        {
            var repo = new InMemoryRepository<User>(u => u.Id);
            var service = new UserService(repo);

            var user = new User("1", "Alice") { Email = "alice@example.com" };
            await repo.SaveAsync(user);

            // Null-conditional and null-coalescing
            var found = await repo.FindAsync("1");
            var name = found?.Name ?? "Unknown";

            // Tuple
            var (id, userName) = user;

            // Using declaration
            await using var stream = new System.IO.MemoryStream();

            Console.WriteLine($"User: {user.ToDisplayString()}");
        }
    }
}

File: initial_exploration/sample_repo/dart/sample.dart
/// Sample Dart file for tree-sitter exploration.
/// Includes Flutter-style patterns and modern Dart features.

import 'dart:async';
import 'dart:convert';

// Type alias
typedef JsonMap = Map<String, dynamic>;
typedef Callback<T> = void Function(T value);

// Enum with enhanced features (Dart 2.17+)
enum Status {
  pending('Pending'),
  active('Active'),
  inactive('Inactive');

  const Status(this.label);
  final String label;

  bool get isTerminal => this == inactive;
}

// Abstract class (interface)
abstract class Repository<T> {
  Future<T?> find(String id);
  Future<void> save(T item);
  Future<bool> delete(String id);
}

// Mixin
mixin Timestamped {
  DateTime? createdAt;
  DateTime? updatedAt;

  void touch() {
    updatedAt = DateTime.now();
  }
}

mixin Identifiable {
  String get id;
}

// Class with mixins
class User with Timestamped, Identifiable {
  @override
  final String id;
  String name;
  String? email;
  Status _status = Status.pending;

  // Constructor with initializer list
  User({
    required this.id,
    required this.name,
    this.email,
  }) : createdAt = DateTime.now();

  // Named constructor
  User.guest() : this(id: 'guest', name: 'Guest User');

  // Factory constructor
  factory User.fromJson(JsonMap json) {
    return User(
      id: json['id'] as String,
      name: json['name'] as String,
      email: json['email'] as String?,
    );
  }

  // Getter
  Status get status => _status;

  // Setter with validation
  set status(Status value) {
    if (_status.isTerminal) {
      throw StateError('Cannot change terminal status');
    }
    _status = value;
    touch();
  }

  // Method
  JsonMap toJson() => {
        'id': id,
        'name': name,
        'email': email,
        'status': status.name,
      };

  // Async method
  Future<void> activate() async {
    await Future.delayed(const Duration(milliseconds: 100));
    status = Status.active;
  }

  // Operator override
  @override
  bool operator ==(Object other) =>
      identical(this, other) || other is User && id == other.id;

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() => 'User(id: $id, name: $name)';
}

// Extension
extension UserExtension on User {
  String get displayName => '$name ($id)';

  bool get hasEmail => email != null && email!.isNotEmpty;
}

// Extension on built-in type
extension StringExtension on String {
  String capitalize() =>
      isEmpty ? this : '${this[0].toUpperCase()}${substring(1)}';
}

// Generic class
class InMemoryRepository<T extends Identifiable> implements Repository<T> {
  final Map<String, T> _items = {};

  @override
  Future<T?> find(String id) async {
    return _items[id];
  }

  @override
  Future<void> save(T item) async {
    _items[item.id] = item;
  }

  @override
  Future<bool> delete(String id) async {
    return _items.remove(id) != null;
  }

  // Stream
  Stream<T> watchAll() async* {
    for (final item in _items.values) {
      yield item;
    }
  }
}

// Abstract class with implementation
abstract class BaseService<T extends Identifiable> {
  final Repository<T> repository;

  BaseService(this.repository);

  Future<T?> getById(String id) => repository.find(id);

  // Abstract method
  bool validate(T item);
}

// Concrete service
class UserService extends BaseService<User> {
  UserService(Repository<User> repository) : super(repository);

  @override
  bool validate(User item) => item.name.isNotEmpty;

  // Null-aware cascade
  Future<User> createUser(String name, {String? email}) async {
    final user = User(
      id: DateTime.now().millisecondsSinceEpoch.toString(),
      name: name,
    )..email = email;

    await repository.save(user);
    return user;
  }
}

// Sealed class (Dart 3.0+)
sealed class Result<T> {
  const Result();
}

class Success<T> extends Result<T> {
  final T value;
  const Success(this.value);
}

class Failure<T> extends Result<T> {
  final String error;
  const Failure(this.error);
}

// Pattern matching (Dart 3.0+)
String handleResult(Result<User> result) {
  return switch (result) {
    Success(value: final user) => 'Found: ${user.name}',
    Failure(error: final msg) => 'Error: $msg',
  };
}

// Record type (Dart 3.0+)
typedef UserRecord = ({String id, String name, String? email});

UserRecord toRecord(User user) => (id: user.id, name: user.name, email: user.email);

// Top-level function
Future<void> main() async {
  final repo = InMemoryRepository<User>();
  final service = UserService(repo);

  // Cascade notation
  final user = await service.createUser('Alice')
    ..email = 'alice@example.com'
    ..status = Status.active;

  print(user.displayName);
  print('Capitalized: ${'hello'.capitalize()}');

  // Collection literals with spread
  final users = [user];
  final moreUsers = [...users, User.guest()];

  // Collection if/for
  final activeUsers = [
    for (final u in moreUsers)
      if (u.status == Status.active) u
  ];

  // Null-aware operators
  final email = user.email ?? 'no-email';
  final length = user.email?.length ?? 0;

  print('Active users: ${activeUsers.length}');
}
